tests/frontend/test_space_unroll.py

```diff
@@ -0,0 +1,250 @@
+
+import pytest
+import numpy as np
+import strawberryfields as sf
+from strawberryfields.tdm.tdmprogram import get_mode_indices
+from strawberryfields.ops import Sgate, Rgate, BSgate, LossChannel, MeasureFock
+from thewalrus.symplectic import reduced_state
+from thewalrus.quantum import is_pure_cov
+
+pytestmark = pytest.mark.frontend
+
+
+def generate_valid_bs_sequence(delays, modes, random_func=np.random.rand):
+    """
+    Generate sequences of valied beamsplitter angles for a time-domain program where
+    the modes inside the delays loops are in vacuum.
+
+    Args:
+        delays (list): list containing the number of modes each loop can support
+        modes (list): number of computational modes to be entangled
+        random_func (function): function to draw samples from
+
+    Returns:
+        array: array containing a valid sequency of beamsplitter that will not entangle the computational modes
+            with the mode inhabiting the loops at the beginning of the sequence
+    """
+    angles = random_func(len(delays), modes + sum(delays))
+    sum_delay = 0
+    for i, delay in enumerate(delays):
+        angles[i, modes + sum_delay :] = 0
+        sum_delay += delay
+        angles[i, :sum_delay] = 0
+    return angles
+
+
+def generate_valid_r_sequence(delays, modes, random_func=np.random.rand):
+    """
+    Generate sequences of valid rotation angles for a time-domain program where
+    the modes inside the delays loops are in vacuum.
+
+    Args:
+        delays (list): list containing the number of modes each loop can support
+        modes (list): number of computational modes to be entangled
+        random_func (function): function to draw samples from
+
+    Returns:
+        array: array containing a valid sequency of rotation angles that will not act like the identity in the modes
+            inhabiting the loops at the beginning of the sequence
+    """
+    angles = random_func(len(delays), modes + sum(delays))
+    sum_delay = delays[0]
+    angles[0, modes:] = 0
+    if len(delays) > 1:
+        for i, delay in enumerate(delays[1:]):
+            angles[i + 1, :sum_delay] = 0
+            angles[i + 1, sum_delay + modes :] = 0
+            sum_delay += delay
+    return angles
+
+
+@pytest.mark.parametrize("delays", [np.random.randint(low=1, high=10, size=i) for i in range(2, 6)])
+@pytest.mark.parametrize("modes", [10, 50, 100])
+def test_lossless_no_mixing_no_rotation_U(delays, modes):
+    """Test that the U matrix is the identity if there is no beamsplitter mixing and no rotations"""
+    delays = list(delays)
+    angles = np.zeros([2 * len(delays), modes + sum(delays)])
+    d = len(delays)
+    n, N = get_mode_indices(delays)
+    prog = sf.TDMProgram([N])
+
+    with prog.context(*angles) as (p, q):
+        for i in range(d):
+            Rgate(p[i + d]) | q[n[i]]
+            BSgate(p[i], np.pi / 2) | (q[n[i + 1]], q[n[i]])
+
+    prog.space_unroll()
+
+    compiled = prog.compile(compiler="passive")
```

Comment:
How about testing `space_unroll` without compiling the program? Could we make sure that
a) The new modes have been added
b) The correct commands have been added to the space unrolled program?